File: set2q8.m
clc;
clear all
close all
im=imread('lena_RGB.tif');
im=double(im);
[r,c,t]=size(im);
R=im(:,:,1);
G=im(:,:,2);
B=im(:,:,3);
ws=5;
cent=(ws+1)/2;
lap=-1*ones(ws,ws);
lap(cent,cent)=-1*sum(lap(:));
pd=(ws-1)/2;
start=ws-pd;
f1=padarray(R,[pd pd],'replicate');
f2=padarray(G,[pd pd],'replicate');
f3=padarray(B,[pd pd],'replicate');

for i=start:r-pd
    for j=start:c-pd
        window1=f1(i-pd:i+pd,j-pd:j+pd);
        window2=f2(i-pd:i+pd,j-pd:j+pd);
        window3=f3(i-pd:i+pd,j-pd:j+pd);
        su1=0; 
        su2=0;  
        su3=0;  
        for s=1:ws
            for t=1:ws
                su1=su1+lap(s,t)*window1(s,t);
                su2=su2+lap(s,t)*window2(s,t);
                su3=su3+lap(s,t)*window3(s,t);
            end
        end
        YR(i-start+1,j-start+1)=su1;
        YG(i-start+1,j-start+1)=su2;
        YB(i-start+1,j-start+1)=su3;
    end
end
Out(:,:,1)=YR;
Out(:,:,2)=YG;
Out(:,:,3)=YB;
figure,subplot(1,2,1),imshow(uint8(im)),title('Original Image');
subplot(1,2,2),imshow(uint8(Out)),title('Sharpened Image');

File: set2q8.py
import numpy as np 
import imageio
from matplotlib import pyplot as plt 
import cv2
def convol(im,kernel):
    ws,ws1=kernel.shape;
    pd=int((ws-1)/2);
    start=ws-pd-1;
    row,col=im.shape
    f=np.pad(im,[(pd,pd)],'reflect')
    r,c=f.shape;
    out=np.zeros([row,col])
    for i in range(start,r-pd):
        for j in range(start,c-pd):
           window=f[(i-pd):(i+pd+1),(j-pd):(j+pd+1)];
           m=np.multiply(kernel,window)
           su=np.sum(np.ravel(m))
           out[i-start,j-start]=su;
    
    return(out)
img = cv2.imread('lena_RGB.tif') 
m,n,o=img.shape
out=np.zeros([m,n,o]) 
R=img[:,:,0]
G=img[:,:,1]
B=img[:,:,2]
ws=3
pd=int((ws-1)/2)
kernel = -1*np.ones((ws,ws))
kernel[pd,pd]=-1*np.sum(np.ravel(kernel))
R1=convol(R,kernel)
G1=convol(G,kernel)
B1=convol(B,kernel)
R1=np.clip(R1,0,255)
G1=np.clip(G1,0,255)
B1=np.clip(B1,0,255)
out[:,:,0]=R1
out[:,:,1]=G1
out[:,:,2]=B1
cv2.imshow('Original Image',img)
cv2.imshow('Sharpened Image',out)
cv2.waitKey(0)
